backend/corpora/common/entities/collection.py

```diff
@@ -174,74 +244,76 @@ def publish(self, data_submission_policy_version):
-
-        # Copy over relationships
-        for link in self.links:
-            link.collection_visibility = CollectionVisibility.PUBLIC
 
         has_dataset_changes = False
         for dataset in self.datasets:
-            revision = Dataset(dataset)
-            original = Dataset.get(self.session, revision.original_id) if revision.original_id else None
+            revised_dataset = Dataset(dataset)
+            original = Dataset.get(self.session, revised_dataset.original_id) if revised_dataset.original_id else None
             if original and public_collection.check_has_dataset(original):
-                dataset_is_changed = original.publish_revision(revision, now)
+                dataset_is_changed = original.publish_revision(revised_dataset, now)
                 if dataset_is_changed:
                     has_dataset_changes = True
             else:
                 # The dataset is new
-                revision.publish_new(now)
+                revised_dataset.publish_new(now)
                 has_dataset_changes = True
 
-        if is_existing_collection and has_dataset_changes:
-            public_collection.update(commit=False, remove_attr="revised_at", revised_at=now)
+        self.session.flush()
+        self.session.expire_all()
+
+        if is_existing_collection:
+            if has_dataset_changes:
+                public_collection.update(commit=False, remove_attr="revised_at", revised_at=now)
+            self.delete()
+            self.db_object = public_collection.db_object
 
         self.session.commit()
-        # commit expires the session, need to retrieve the original private collection to delete it
-        private_collection = Collection.get_collection(self.session, self.id, CollectionVisibility.PRIVATE)
-        private_collection.delete()
-        self.db_object = public_collection.db_object
 
-    def revision(self) -> "Collection":
+    def create_revision(self) -> "Collection":
         """
         Generate a collection revision from a public collection
         :return: collection revision.
 
         """
         revision_collection = clone(
-            self.db_object, primary_key=dict(id=self.id, visibility=CollectionVisibility.PRIVATE)
+            self.db_object,
+            primary_key=dict(id=generate_uuid()),
+            visibility=CollectionVisibility.PRIVATE,
+            revision_of=self.id,
         )
         self.session.add(revision_collection)
         for link in self.links:
-            self.session.add(clone(link, collection_id=self.id, collection_visibility=CollectionVisibility.PRIVATE))
-        self.session.commit()
+            self.session.add(clone(link, collection_id=revision_collection.id))
         for dataset in self.datasets:
-            Dataset(dataset).create_revision()
+            Dataset(dataset).create_revision(revision_collection.id)
+        self.session.commit()  # why is this happening here and not after creating the Dataset revisions
```

Comment:
rm b/c `Dataset(dataset).create_revision(revision_collection.id)` already calls commit.

Reply:
Is that bad practice? Asking honestly 🤔

Reply:
it is bad practice to be using commits embedded in lower functions. We want to keep the transactions as tight as possible. At most, we should be doing a flush, but I was young and naive when I originally implemented it. If there is nothing to commit sqlalchemy will thrown an exception.

Reply:
Right, I agree that we shouldn't have the commits in the lower functions, but those were already there, and I wasn't interested in doing a refactor for that work, which would be much more involved.